### PR TITLE
Be more efficient in how we look up watched WS paths.

### PR DESCRIPTION
**Pull request by brian-brazil**
1.5s -> 60ms.

### CODE REVIEW

**Comment by brian-brazil** [comment]
Hold up, might be a bug here.

**Comment by brian-brazil** [comment]
Okay, expanded the scope a bit. We're now going from ~20s to ~1.5s overall.

**Comment by brian-brazil** [comment]
And PT loading with 90 penalties is now 5x faster. Creating penalties is already quite fast, so no need for changes there.

File: src/com/carolinarollergirls/scoreboard/jetty/WS.java [hunk 1]
```diff
@@ -266,7 +265,54 @@ private void sendWSUpdatesForPaths(Set<String> watchedPaths, Set<String> changed
...
+          }
+          public void add(String p) {
+            PathTrie head = this;
+            for (int i = 0;; i++) {
+              if (head.exists) {
+                // Already covered.
+                return;
+              }
+              if (i >= p.length()) {
+                break;
```
Comment (frank-weinberg) [inline src/com/carolinarollergirls/scoreboard/jetty/WS.java:288]: This condition should be in the head of the loop.

File: src/com/carolinarollergirls/scoreboard/jetty/WS.java [hunk 2]
```diff
@@ -266,7 +265,54 @@ private void sendWSUpdatesForPaths(Set<String> watchedPaths, Set<String> changed
...
+                 head = child;
+              }
+            }
+            head.exists = true;
+          }
+          public boolean covers(String p) {
+            PathTrie head = this;
+            for (int i = 0;; i++) {
+              if (head.exists) {
+                // Already covered.
```
Comment (frank-weinberg) [inline src/com/carolinarollergirls/scoreboard/jetty/WS.java]: This comment doesn't really make sense in this function.

File: html/controls/sk/sk-sheet.js [hunk 3]
```diff
@@ -139,11 +127,11 @@ function prepareSkSheetTable(element, teamId, mode) {
...
 				otherRow.find('.Trip2').text('');
-			});
-			$.each(new Array(7), function (idx) {
-				var t = idx + 3;
-				WS.Register([prefix+'ScoringTrip('+t+').Score', prefix+'ScoringTrip('+t+').Current', 
-						prefix+'ScoringTrip('+t+').AfterSP'], function (k, v) {
+				break;
+
+			 default:
+				if (k.parts[4] == 'ScoringTrip' && k.ScoringTrip >= 3 && k.ScoringTrip <= 10) {
```
Comment (frank-weinberg) [inline html/controls/sk/sk-sheet.js]: This should be <10.
Trip 10 is already part of the cell that's treated below.

File: html/controls/sk/sk-sheet.js [hunk 4]
```diff
@@ -154,48 +142,95 @@ function prepareSkSheetTable(element, teamId, mode) {
...
-				var scoreAfterSP = '';
-				$.each(new Array(3), function(idx) {
-					var t = idx + 10;
-					var tripScore = WS.state[prefix+'ScoringTrip('+t+').Score'];
-					if (tripScore != null) {
-						if (isTrue(WS.state[prefix+'ScoringTrip('+t+').AfterSP'])) {
-							scoreAfterSP = scoreAfterSP=='' ? tripScore : scoreAfterSP + "+" + tripScore;
-						} else {
-							scoreBeforeSP = scoreBeforeSP=='' ? tripScore : scoreBeforeSP + "+" + tripScore;
+				} else if (k.parts[4] == 'ScoringTrip' && k.ScoringTrip > 10) {
```
Comment (frank-weinberg) [inline html/controls/sk/sk-sheet.js]: `>= 10`

File: html/controls/sk/sk-sheet.js [hunk 5]
```diff
@@ -154,48 +142,95 @@ function prepareSkSheetTable(element, teamId, mode) {
...
+					var scoreBeforeSP = '';
+					var scoreAfterSP = '';
+					$.each(new Array(3), function(idx) {
+						var t = idx + 10;
+						var tripScore = WS.state[prefix+'ScoringTrip('+t+').Score'];
+						if (tripScore != null) {
+							if (isTrue(WS.state[prefix+'ScoringTrip('+t+').AfterSP'])) {
+								scoreAfterSP = scoreAfterSP=='' ? tripScore : scoreAfterSP + "+" + tripScore;
+							} else {
+								scoreBeforeSP = scoreBeforeSP=='' ? tripScore : scoreBeforeSP + "+" + tripScore;
```
Comment (frank-weinberg) [inline html/controls/sk/sk-sheet.js]: These need spaces around the `+` so the text will wrap when needed

File: html/controls/sk/sk-sheet.js [hunk 6]
```diff
@@ -154,48 +142,95 @@ function prepareSkSheetTable(element, teamId, mode) {
...
+							}
 						}
-					}
-				});
-				jamRow.find('Trip10').text(scoreBeforeSP);
-				spRow.find('Trip10').text(scoreAfterSP);
+					});
+					jamRow.find('Trip10').text(scoreBeforeSP);
+					spRow.find('Trip10').text(scoreAfterSP);
```
Comment (frank-weinberg) [inline html/controls/sk/sk-sheet.js]: These need to be `.Trip10`

File: html/controls/sk/sk-sheet.js [hunk 7]
```diff
@@ -154,48 +142,95 @@ function prepareSkSheetTable(element, teamId, mode) {
...
-					var tripScore = WS.state[prefix+'ScoringTrip('+t+').Score'];
-					if (tripScore != null) {
-						if (isTrue(WS.state[prefix+'ScoringTrip('+t+').AfterSP'])) {
-							scoreAfterSP = scoreAfterSP=='' ? tripScore : scoreAfterSP + "+" + tripScore;
-						} else {
-							scoreBeforeSP = scoreBeforeSP=='' ? tripScore : scoreBeforeSP + "+" + tripScore;
+				} else if (k.parts[4] == 'ScoringTrip' && k.ScoringTrip > 10) {
+					var scoreBeforeSP = '';
+					var scoreAfterSP = '';
+					$.each(new Array(3), function(idx) {
```
Comment (frank-weinberg) [inline html/controls/sk/sk-sheet.js]: Since you're now catching any trip index, this should be a while loop that ends, when a trip score is `null`.

File: html/controls/sk/sk-sheet.js [hunk 8]
```diff
@@ -154,48 +142,95 @@ function prepareSkSheetTable(element, teamId, mode) {
...
 				table.prepend(jamRow).prepend(spRow);
 			} else {
 				table.append(jamRow).append(spRow);
 			}
+			jamElements[p][nr] = [jamRow, spRow];
+
 		}
 	}
-	
-	function finalizePeriod(nr) {} //TODO: implement
```
Comment (frank-weinberg) [inline html/controls/sk/sk-sheet.js:198]: Why did you remove this? This is still an active todo on my list.

File: html/controls/plt/plt-input.js [hunk 9]
```diff
@@ -29,7 +29,11 @@ function preparePltInputTable(element, teamId, mode, statsbookPeriod, alternateN
 			$('<td>').text('Pivot').appendTo(thead);
 			$('<td>').text('Blocker').appendTo(thead);
 			$('<td>').text('Box').appendTo(thead);
-			$('<td>').appendTo(thead);
+			if (mode == 'lt') {
+				$('<td>').attr('id', 'head').text('Team ' + teamId).appendTo(thead);
```
Comment (frank-weinberg) [inline html/controls/plt/plt-input.js:33]: This field is to small for a long team name.
I'd rather use it for the SP indicator (on PLT as well).

**Review by frank-weinberg — COMMENTED** [review]
The period and jam tables in the operator screen could probably also profit from a rework using the amended WS keys.

File: src/com/carolinarollergirls/scoreboard/jetty/WS.java (hunk 1 above)
Comment (brian-brazil) [reply src/com/carolinarollergirls/scoreboard/jetty/WS.java:288]: That'd result in us iterating once when we don't need to, which is why the head.exists check is before it.

File: src/com/carolinarollergirls/scoreboard/jetty/WS.java (hunk 2 above)
Comment (brian-brazil) [reply src/com/carolinarollergirls/scoreboard/jetty/WS.java]: Removed

File: html/controls/sk/sk-sheet.js (hunk 3 above)
Comment (brian-brazil) [reply html/controls/sk/sk-sheet.js]: Changed

File: html/controls/sk/sk-sheet.js (hunk 6 above)
Comment (brian-brazil) [reply html/controls/sk/sk-sheet.js]: Fixed

File: html/controls/sk/sk-sheet.js (hunk 7 above)
Comment (brian-brazil) [reply html/controls/sk/sk-sheet.js]: Done

File: html/controls/sk/sk-sheet.js (hunk 8 above)
Comment (brian-brazil) [reply html/controls/sk/sk-sheet.js:198]: It's dead code currently, it can be added back when needed.

File: html/controls/plt/plt-input.js (hunk 9 above)
Comment (brian-brazil) [reply html/controls/plt/plt-input.js:33]: You're probably right, but it'll do the job for now.

File: html/controls/sk/sk-sheet.js (hunk 5 above)
Comment (brian-brazil) [reply html/controls/sk/sk-sheet.js]: Done, though the wrapping means the SK tables no longer align in the (rare) case someone somehow manages to have 11+ scoring passes.

File: html/controls/sk/sk-sheet.js (hunk 4 above)
Comment (brian-brazil) [reply html/controls/sk/sk-sheet.js]: Done

**Comment by brian-brazil** [comment]
> The period and jam tables in the operator screen could probably also profit from a rework using the amended WS keys.

They're actually a notable hotspot (though this PR helps), so I'll be reworking them at some point.
